covsirphy/cleaning/vaccine_data.py

```diff
@@ -112,6 +113,22 @@ def _cleaning(self):
         for col in [self.VAC, self.V_ONCE, self.V_FULL]:
             df[col] = pd.to_numeric(df[col], errors="coerce")
             df[col] = df.groupby(self.ISO3)[col].fillna(method="ffill").fillna(0).astype(np.int64)
+        for country in df.Country.unique().tolist():
+            subset_df = df.loc[df[self.COUNTRY] == country]
+            # add any missing dates up until today
+            # while keeping the data same from the last date
+            today_date = datetime.datetime.today().replace(hour=00, minute=00, second=00, microsecond=00)
+            last_country_date = subset_df.iloc[-1][self.DATE]
```

Comment:
Thank you for your pull request!!
Please consider the following technic to avoid `while` loop for speed.
```Python
new_dates = pd.date_range(subset_df["Date"].max() + timedelta(days=1)
new_df = pd.DataFrame(index=new_dates, today_date), columns=subset_df.columns)
new_df.index.name = "Date"
new_df = new_df.drop("Date", axis=1).reset_index()
pd.concat([subset_df, new_df], axis=0, ignore_index=True).ffill()
```

Reply:
For some reason the line `new_df = pd.DataFrame(index=new_dates, today_date, columns=subset_df.columns)` gives me unspecified `SyntaxError: invalid syntax`.

Reply:
Sorry, updated my comment to fix SyntaxError.

Reply:
In this case though, we have to assign the `pd.concat()` to `subset_df` but then we have to insert to `df` only the new values (I guess you used concat with ffill in order to get the values and to avoid while loop).
So then we have to keep only the new rows that were inserted to `subset_df` and then append the result to df, like this:
```
        # Fill in NA values
        for col in [self.VAC, self.V_ONCE, self.V_FULL]:
            df[col] = pd.to_numeric(df[col], errors="coerce")
            df[col] = df.groupby(self.ISO3)[col].fillna(method="ffill").fillna(0).astype(np.int64)
        today_date = datetime.datetime.today().replace(hour=00, minute=00, second=00, microsecond=00)
        for country in df.Country.unique():
            subset_df = df.loc[df[self.COUNTRY] == country]
            # add any missing dates up until today
            # while keeping the data same from the last date
            if subset_df["Date"].max() < today_date:
                new_dates = pd.date_range(subset_df["Date"].max() + datetime.timedelta(days=1), today_date)
                keep_index = subset_df["Date"].idxmax()+1
                new_df = pd.DataFrame(index=new_dates, columns=subset_df.columns)
                new_df.index.name = "Date"
                new_df = new_df.drop("Date", axis=1).reset_index()
                subset_df = pd.concat([subset_df, new_df], axis=0, ignore_index=True).ffill()
                subset_df = subset_df.loc[keep_index:]
                df = pd.concat([df, subset_df], axis=0, ignore_index=True)
        df.sort_values(by=[self.COUNTRY, self.DATE], ignore_index=True, inplace=True)
        return df.loc[:, self.VAC_COLS]
```
I don't know, it is more or less the same with the initial commit, except for the while loop. Which one do you prefer?

Reply:
I have checked-in this implementation (without the while loop) so if it is okay we can merge this.

Reply:
Thank you for updating! I will replace `"Date"` with `self.DATE` and merge this pull request.
As much as possible, please avoid combination of pandas & loops for performance.
(We can remove `keep_index` lines if we use `df.drop_duplicates(subset=["Country", "Date"], keep="first")`.)